torchtext/csrc/vectors.cpp

```diff
@@ -60,16 +81,195 @@ struct Vectors : torch::CustomClassHolder {
     if (item != stovec_.end()) {
       item->value() = vector;
     } else {
-      tokens_.push_back(token);
-      vectors_ = torch::cat({vectors_, torch::unsqueeze(vector, /*dim=*/0)},
-                            /*dim=*/0);
-      stovec_.insert_or_assign(token, vectors_.select(0, stovec_.size()));
+      stoindex_.insert_or_assign(token, vector.size(0));
+      stovec_.insert_or_assign(token, vector);
+      // TODO: This could be done lazily during serialization (if necessary).
+      // We would cycle through the vectors and concatenate those that aren't
+      // views.
+      vectors_ = at::cat({vectors_, vector.unsqueeze(0)});
     }
   }
 
   int64_t __len__() { return stovec_.size(); }
 };
 
+inline int64_t divup(int64_t x, int64_t y) { return (x + y - 1) / y; }
+
+std::tuple<int64_t, int64_t, int64_t>
+_infer_shape(const std::string &file_path, const int64_t delimiter_ascii) {
+
+  int64_t num_header_lines = 0, num_lines = 0, vector_dim = -1;
+  std::vector<std::string> vec_str;
+  std::string line, word;
+
+  std::ifstream fin;
+  fin.open(file_path, std::ios::in);
+
+  while (std::getline(fin, line)) {
+    vec_str.clear();
+    if (vector_dim == -1) {
+      std::istringstream s(line);
+
+      // get rid of the token
+      std::getline(s, word, static_cast<char>(delimiter_ascii));
+      // we assume entries for vector are always seperated by ' '
+      while (std::getline(s, word, ' ')) {
+        vec_str.push_back(word);
+      }
+
+      // assuming word, [vector] format
+      // the header present in some(w2v) formats contains two elements
+      if (vec_str.size() <= 2) {
+        num_header_lines++;
+      } else if (vec_str.size() > 2) {
+        vector_dim = vec_str.size();
+        num_lines++; // first element read
+      }
+    } else {
+      num_lines++;
+    }
+  }
+  return std::make_tuple(num_lines, num_header_lines, vector_dim);
+}
+
+void _infer_offsets(const std::string &file_path, int64_t num_lines,
+                    int64_t chunk_size, int64_t num_header_lines,
+                    std::vector<size_t> *offsets) {
```

Comment:
Why not pass by reference?

Reply:
I was just trying to adhere to Google's C++ standard,

> In fact it is a very strong convention in Google code that input arguments are values or const references while output arguments are pointers.

https://drake.mit.edu/styleguide/cppguide.html#Reference_Arguments

Reply:
> I was just trying to adhere to Google's C++ standard,
> 
> > In fact it is a very strong convention in Google code that input arguments are values or const references while output arguments are pointers.
> 
> https://drake.mit.edu/styleguide/cppguide.html#Reference_Arguments

I do not see the corresponding one in the official style guide.

https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

Reply:
You're right maybe they changed the official guide since the last time I read over it! Updated to use pass by reference.